src/translate.cpp

```diff
@@ -0,0 +1,105 @@
+/*
+ *  gd-tools - a set of programs to enhance goldendict for immersion learning.
+ *  Copyright (C) 2025 Ajatt-Tools
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "precompiled.h"
+#include "util.h"
+
+using namespace std::literals;
+namespace sp = subprocess;
+
+static constexpr std::string default_to{ "en" };
+static constexpr std::string_view help_text = R"EOF(usage: gd-translate [OPTIONS]
+
+Translate text from Japanese to target language
+
+OPTIONS
+        --to LANG             target language (default: en)
+  --sentence SENTENCE         japanese sentence to translate
+   --spoiler yes/no           black out the sentence with a spoiler box (default: no)
+
+EXAMPLES
+  gd-translate --spoiler yes --sentence %GDSEARCH%
+  gd-translate --spoiler yes --to fr --sentence %GDSEARCH%
+)EOF";
+static constexpr std::string_view css_style = R"EOF(<style>
+    .spoiler {
+      background-color: black;
+      padding: 6px;
+      width: fit-content;
+    }
+    .spoiler:hover {
+      background-color: white;
+    }
+</style>)EOF";
+
+struct translate_params
+{
+  std::string to{ default_to };
+  std::string gd_word;
+  std::string spoiler{ "no" };
+
+  void assign(std::string_view const key, std::string_view const value)
+  {
+    if (key == "--to") {
+      to = value;
+    } else if (key == "--sentence") {
+      gd_word = value;
+    } else if (key == "--spoiler") {
+      if (value == "yes") {
+        spoiler = value;
+      }
+    }
+  }
+};
+
+void exec_translate(translate_params const& params)
+{
+  auto cmd_argos = sp::Popen(
+    {
+      "argos-translate",
+      "-f",
+      "ja",
+      "-t",
+      params.to,
+      params.gd_word,
+    },
+    sp::output{ sp::PIPE }
+  );
+
+  auto cmd_tail = sp::Popen({ "tail", "-n1" }, sp::input{ cmd_argos.output() }, sp::output{ sp::PIPE });
```

Comment:
Do you actually need `tail -n1` here? As far as I see, `argos-translate` sends the translation to stdout, and everything else to stderr. I'd just suppress stderr and print stdout as is.

Reply:
You're right, it is going to stderr.